docs/papi.1
.\" Automatically generated by Pandoc 1.19.2.4
.\"
.TH "Papi" "1" "" "Version 0.99" "Parsing and printing interpreter"
.hy
.SH NAME
.PP
\f[B]papi\f[] \- interpreter for user\-defined parsers
.SH SYNOPSIS
.PP
\f[B]papi\f[] [options] \f[I]grammar\-file\f[] \f[I]input\-file\f[]
.SS Options
.TP
.B \-d
Debug mode, dumps parser stack even after parsing success
.RS
.RE
.TP
.B \-g
Prints user grammar as JSON array
.RS
.RE
.TP
.B \-h
Prints usage information
.RS
.RE
.TP
.B \-j
Forces result output format to be JSON array
.RS
.RE
.TP
.B \-m
Prints the papi metalanguage grammar in the form of limited JSON
.RS
.RE
.TP
.B \-r
Prints the result in the form of a raw string
.RS
.RE
.TP
.B \-s
Silent mode \- do not dump parser stack after failure
.RS
.RE
.SH DESCRIPTION
.PP
Papi (Parsing And Printing Interpreter) is an interpreter of parsers for
user\-defined languages.
Grammar of user language is defined using notation based on PEG (Parsing
Expression Grammars).
The notation is hereinafter referred to as the "papi metalanguage".
.PP
The command \f[B]papi grammar\-file input\-file\f[] reads definition of
user language from \f[I]grammar\-file\f[].
It is parsed by built\-in parser of the papi metalanguage.
.PP
If this step is successful, the program reads text from
\f[I]input\-file\f[], parses it with the user parser obtained in the
previous step, and if this step is also successful, the result will be
printed on the output.
How it will be printed depends on the type of result and program
options.
If the result is a single string (and not a sequence or list) or when
the \f[B]\-r\f[] option is used, it is treated as a raw string and
printed directly on the output.
In turn, if it is not a string, or the \f[B]\-j\f[] option is used, the
result is printed using JSON notation, limiting the notation to tables
and strings.
.PP
If any of the steps fail, papi dumps the stack of parser that failed
(metalanguage parser or user language parser), saved when the parser
reached the position farthest to the right.
.SS Structure and operation of the papi parser
.PP
Both the parser of user grammar and the user language parser used in the
second phase are based on the same mechanism.
In fact, the grammar of the metalanguage is also written in the
metalanguage itself.
It is available in the file "metagrammar.dp" in the program source.
The line numbers used in the metalanguage parser error messages refer to
this file.
.PP
The papi parser is based on modified PEG (Parsing Expression Grammars)
formalism.
The original PEG formalism does not specify mechanisms for defining and
passing data which is a result of parser operations.
For that purpose, papi utilises variables (written as identifiers
prefixed with "$") and templates describing the form of results.
.PP
Hence, the metalanguage uses two types of identifiers: production names
(written as a sequence of digits and letters, starting from a letter)
and variable names (starting from the "$").
.SS Data types
.PP
The data object in papi can be a string or a list, with the elements of
the list being objects.
.PP
The parser, as understood by papi, is a procedure which reads input text
data, accepts them \- and then it produces some result, or rejects them
\- and then there is no result.
.PP
The result of parser procedure is a sequence of objects \- zero, one or
more objects \- the sequence hence can be empty.
One should also note that the sequence is not the same as the list.
The list is a single object, with elements inside, and the sequence is a
zero, one or more objects.
.PP
Lists and sequences in the metalanguage are denoted in different ways:
.IP
.nf
\f[C]
\ \ \ Sequence\ \ :\ \ \ \ \ \ \ {object1,\ object2,\ ...}
\ \ \ List\ \ \ \ \ \ :\ \ \ \ \ \ \ [object1,\ object2,\ ...]
\f[]
.fi
.PP
The difference between them is that the sequence is not an object, so it
can be an element of another sequence nor some list.
If it is appended, it means that all its elements are appended in turn:
.IP
.nf
\f[C]
\ \ \ {\ "a",\ {"b",\ "c"}\ }\ =\ {\ "a",\ "b",\ "c"\ }
\ \ \ [\ "a",\ {"b",\ "c"}\ ]\ =\ [\ "a",\ "b",\ "c"\ ]
\f[]
.fi
.PP
The same rule does not apply to lists, because the list is a single
object that is indivisible:
.IP
.nf
\f[C]
\ \ \ [\ "a",\ ["b",\ "c"]\ ]\ =\ [\ "a",\ ["b",\ "c"]\ ]
\ \ \ {\ "a",\ ["b",\ "c"]\ }\ =\ {\ "a",\ ["b",\ "c"]\ }
\f[]
.fi
.SS Structure of grammar
.PP
The grammar in papi consist of productions.
Each production has a unique name and describes a parser.
Production called "goal" is an entry point to the grammar and describes
a parser which is the final product of parsing the grammar, i.e.
the user parser.
.PP
The production consists of the name and one or more branches.
During execution of the procedure described by the production the parser
tries to pass full branch, one after another, until one of them is
finished with success.
.PP
During execution of the procedure described by the production the parser
tries to finish one of the branches, one after another, until one of
them is finished with success.
When all branches fail, the whole production procedure is considered
failed \- then there is no result.
.PP
The branches in the metalanguage code are separated by "|".
This corresponds to the operator of the alternative in PEG formalism.
.PP
Here is an example production with two branches (note the comments
starting with "#"):
.IP
.nf
\f[C]
\ \ \ \ alnum\ =\ letter\ $c\ \->\ $c\ \ \ \ \ \ \ #\ branch\ 1
\ \ \ \ \ \ \ \ \ \ |\ digit\ $c\ \->\ $c\ \ \ \ \ \ \ \ #\ branch\ 2
\f[]
.fi
.SS Repetition operators
.PP
The branch of production consists of a series of parsing elements
optionally equipped with a variable name.
Repetition operators are used to mark fragments of these series to be
repeated zero, one or more times.
There are three kind of repetition operators, known from the ENBF
notation, from regular expressions and from PEG:
.IP
.nf
\f[C]
\ \ \ \ (\ ...\ )?\ \ \ \ \-\ repeat\ zero\ or\ one\ time\ \-\ in\ other\ words\ \-\ "optionally"
\ \ \ \ (\ ...\ )*\ \ \ \ \-\ repeat\ zero\ or\ more\ times
\ \ \ \ (\ ...\ )+\ \ \ \ \-\ repeat\ one\ or\ more\ times
\f[]
.fi
.PP
The repetition operators repeat its component parsers in a loop doing
this as many times as it is possible.
If this succeeds, they collect the results in the variables which were
declared inside.
In effect, each variable collects the sequences of results repeated as
many times as the expression succeeds.
When used in the template, the sequences are restored with the same
number of elements as collected in the parsing phase.
.SS Prefix operators
.PP
In the metalanguage (similarly as in PEG) we have also prefix operators.
They are the negation operator "!" and the "&" predicate tag.
The negation and the predicate are described below when discussing
parsing elements.
.SS Templates
.PP
What is to the right of the "\->" arrow symbol is the template.
It is an expression consisting of variables (those that collect data
during parsing) and constants (strings written as in C or javascript),
possibly enclosed in repetition operators that correspond to analogous
ones on the left side of the production.
.PP
The repetition operators determine also the way in which the sequence of
variable values is joined.
The way is selected depending on the type od used brackets:
.IP
.nf
\f[C]
\ \ \ \ (\ $a,\ "x",\ ...\ )*\ \ \-\ connecting\ to\ string
\ \ \ \ {\ $a,\ "x",\ ...\ }*\ \ \-\ combining\ into\ a\ sequence
\ \ \ \ [\ $a,\ "x",\ ...\ ]*\ \ \-\ joining\ into\ a\ list
\f[]
.fi
.PP
The variables in templates are also used in the scope of repetition
operators "+" and "?", and prefix predicate tag "&":
.IP
.nf
\f[C]
\ \ \ \ ($a,\ "x")*
\ \ \ \ ($a,\ "x")+
\ \ \ \ ($a,\ "x")?
\ \ \ \ &($a,\ "x")
\f[]
.fi
.PP
There is no need to use the negation operator "!" in the templates
because negation produces no result.
.SS Parsing elements
.PP
Here are the types of parsing elements that are defined in the papi
metalanguage:
.TP
.B Call of production
Call of production is written as the called production name with
optional variable name.
The variable is a destination of the sequence of results.
For example:
.RS
.IP
.nf
\f[C]
expr\ $x
\f[]
.fi
.RE
.TP
.B Constant string
The constant is written as in "C" or javascript.
The execution of such element succeeds when the parser manages to read
from input exactly the characters from the string.
The variable name may be also optionally given after the string.
Inside the literal one can use backslashed escape sequences "\\n",
"\\r", "\\t" and "\\"", "\\\\".
Such a details may be found in the definition of the metalanguage
grammar in the file "metagrammarp.dp".
Example:
.RS
.IP
.nf
\f[C]
"The\ end\\\\n"\ \ $footer
\f[]
.fi
.RE
.TP
.B Any character
This basic parser accepts any character from input and fails at the end
of the input file.
Example:
.RS
.IP
.nf
\f[C]
\&.\ $ch
\f[]
.fi
.RE
.TP
.B Character class
List of characters enclosed by "[" and "]".
Accepts any single character from the set.
Some characters have to be escaped by backslash: "\\[", "\\]", "\\\\".
One can also use the "\\r" and "\\n" sequences.
Example:
.RS
.IP
.nf
\f[C]
[0123456789]\ $digit
\f[]
.fi
.RE
.TP
.B Predicate
Sequence of elements in parentheses preceded by "&".
The predicate executes all its components in turn and makes their
results available (using the "&" operator in the template), but finally
moves the reading pointer back to the starting position.
This is a notion derived from PEG.
It extends the meaning of the grammar class that can be created using
this notation.
For example:
.RS
.IP
.nf
\f[C]
&(name\ $n\ value\ $x)
\f[]
.fi
.RE
.TP
.B Negation
Using the negation operator ("!") before an element we get the parser
which works in a way opposite to the original one.
At first, it tries to parse the data with its inner element, and then it
inverts the decision \- if the parsing succeeds it reverts the read
pointer to the starting position and returns the false.
If not, it returns the success (the negation succeeds).
The negation never has meaningful data associated with its result, so
the variable cannot be used.
.RS
.RE
.SH FILES
.PP
\f[I]metagrammar.dp\f[] \- Description of the language of metagramatics
(metalanguage) written in the metalanguage itself.
This file is used in booststrap process to generate the parser of
metagrammar in papi.
.PP
The current version does not use any configuration files nor environment
variables.


src/papi.cc
#include  "grammar.h"

#include  <cstdlib>
#include  <cstring>
#include  <fstream>        
#include  <iostream>
#include  <unistd.h>

#define  PROGRAM_VERSION  "0.99"

static const char *banner = "papi  " PROGRAM_VERSION "\n";

static const char *usage_string =
"Usage   :  papi [options] grammar_file data_file\n"
"Options:  -d    - debug mode, always dump parser stack\n"
"          -g    - dump user grammar (in json list format)\n"
"          -h    - help\n"
"          -j    - result output format = json lists\n"
"          -m    - dump metagrammar (in json list format)\n"
"          -r    - result output format = raw string result\n"
"          -s    - silent mode - do not dump parser stack on error\n";

static  void  show_usage()
{
        std::cout << banner << usage_string;
}

/*  Output modes  */
enum class OutputMode {
        Undefined,   //  to be recognized based on data structure 
        JsonList,    //  using subset of JSON (strings and lists)
        Text         //  as raw string text
};

/*  Command line options values  */
bool  op_debug = false;
bool  op_dump_grammar = false;
OutputMode  op_output_mode = OutputMode::Undefined;
bool  op_silent = false;

std::string  grammar_filename, data_filename;
bool  action_performed = false;

OutputMode  recognize_output_mode( const std::vector<LisData> &data )
{
        if ( data.size() == 1 )
        {
                try {
                        data.at( 0 ).getString();
                        return  OutputMode::Text;
                }
                catch ( const std::bad_variant_access& )
                {
                }
        }

        return  OutputMode::JsonList;
}

std::string  json_escape_string( const std::string &s )
{
        std::string  es = "\"";

        for ( auto ch: s )
                switch( ch )
                {
                        case '\r':
                                es += "\\r";
                                break;
                        case '\n':
                                es += "\\n";
                                break;
                        case '\t':
                                es += "\\t";
                                break;
                        case '\"':
                                es += "\\\"";
                                break;
                        case '\\':
                                es += "\\\\";
                                break;
                        default:
                                if ( isprint(ch) )
                                        es += ch;
                                else
                                {
                                        char xbuf[ 8 ];
                                        snprintf( xbuf, sizeof(xbuf), "\\x%02x", int(ch) );
                                        es += xbuf;
                                }
                }

        return es + "\"";
}

std::string  output_as_string( const std::vector<LisData> &data )
{
        std::string  s;

        for ( const auto it: data )
        {
                try {
                        s += it.getString();
                }
                catch ( const std::bad_variant_access& )
                {
                }
        }

        return s;
}

std::string  output_as_json_list( const std::vector<LisData> &data )
{
        std::string  s = "[";
        std::string  sep;

        for ( const auto it: data )
        {
                try {
                        s += sep + json_escape_string( it.getString() );
                }
                catch ( const std::bad_variant_access& )
                {
                        s += sep + output_as_json_list( it.getList() );
                }

                sep = ",";
        }

        return s + "]";
}

std::string  output_as_json_sequence( const std::vector<LisData> &data )
{
        std::string  s, sep;
        bool  numbers = true;

        if ( data.size() == 1 )
                numbers = false;

        int  number = 1;

        for ( const auto it: data )
        {
                s += sep;

                if ( numbers )
                        s += std::to_string( number++ ) + ": ";

                try {
                        s += json_escape_string( it.getString() );
                }
                catch ( const std::bad_variant_access& )
                {
                        s += output_as_json_list( it.getList() );
                }

                sep = "\n";
        }

        return s;
}

std::string  output_result( OutputMode  mode,  const std::vector<LisData> &data )
{
        switch ( mode ) 
        {
                case  OutputMode::Text:
                        return output_as_string( data );
                case  OutputMode::JsonList:
                        return output_as_json_sequence( data );
                default:
                        mode = recognize_output_mode( data );
                        return  output_result( mode, data );
        }
}

void  dump_parser_stack( ParserState<Info> *ps, const std::string &filename, 
                const std::string &message )
{
        std::cout << message << ",\nParser stack:\n";

        std::string  dump;

        for ( auto it: ps->getInfo() )
        {
                dump = it.first + ", at '" + filename + 
                        "' in: " + std::to_string( it.second.pos ) + 
                        " [line " + std::to_string( it.second.line ) + ": column " + 
                        std::to_string( it.second.column ) + ']' + '\n' + dump;
        }

        std::cout << dump;
}

int  main( int argc, char *argv[] )
{
#include  "metagenerated.h"
        LisData  metagrammarStruct = metagenerated;

        int op = 0;

        while ( op >= 0 )
        {
                op = getopt( argc, argv, "dhjgmrs" );
                switch ( op )
                {
                        case 'd':
                                op_debug = true;
                                break;
                        case 'g':
                                op_dump_grammar = true;
                                break;
                        case 'h':
                                show_usage();
                                action_performed = true;
                                std::exit( 0 );
                        case 'j':
                                op_output_mode = OutputMode::JsonList;
                                break;
                        case 'm':
                                std::cout << 
                                        output_as_json_sequence( {metagrammarStruct} ) << '\n';

                                action_performed = true;
                                break;
                        case 'r':
                                op_output_mode = OutputMode::Text;
                                break;
                        case 's':
                                op_silent = true;
                                break;
                        case -1:
                                break;
                        default:
                                exit( 1 );
                }
        }

        if ( optind < argc )
                grammar_filename = argv[ optind++ ];

        if ( optind < argc )
                data_filename = argv[ optind++ ];

        try {
                LisData   userGrammarStruct;
                GMap      metagrammar, userGrammar;      
                std::vector<LisData>   userData;  

                if ( grammar_filename.empty() )
                {
                        if ( !action_performed ) 
                                show_usage();

                        std::exit( 0 );
                }

                compileGrammar( metagrammar, metagrammarStruct, 
                                "internal metagrammar", grammar_filename.c_str() );

                if ( !metagrammar.count( "goal" ))
                {
                        std::cout << "Internal error - no 'goal' in the metagrammar\n";
                        std::exit( 2 );
                }

                ParserState<Info> *ps = nullptr;
                ps = new ParserState<Info>( new std::ifstream(grammar_filename) );

                Parser<Info, std::vector<LisData>>  p = *metagrammar.at("goal");
                std::optional<std::vector<LisData>>  r = p( *ps );

                if ( r ) 
                {
                        if ( r->size() != 1 )
                                throw  Exception( grammar_filename + 
                                        ": internal problem - vector parsed has " + 
                                        std::to_string( r->size() ) +
                                        " elements" );

                        userGrammarStruct = r->at( 0 );
                        
                        if ( op_debug )
                                dump_parser_stack( ps, grammar_filename, 
                                                "Metagrammar parser succeeded" );

                        if ( op_dump_grammar )        
                                std::cout << 
                                        output_as_json_sequence( {userGrammarStruct} ) << '\n';

                }
                else
                {
                        if ( op_silent )
                                std::cout << "Metagrammar parser failed\n";
                        else
                                dump_parser_stack( ps, grammar_filename, 
                                                "Metagrammar parser failed" );

                        std::exit( 1 );
                }

                compileGrammar( userGrammar, userGrammarStruct, 
                                grammar_filename.c_str(), data_filename.c_str() );

                if ( !( data_filename.empty() || userGrammar.count( "goal" )))
                {
                        std::cout << grammar_filename << ": no 'goal' in the grammar\n";
                        std::exit( 2 );
                }

                if ( data_filename.empty() ) 
                        std::exit( 0 );

                Parser<Info, std::vector<LisData>>  *ptr = nullptr;
                ptr = userGrammar.at( "goal" );

                delete ps;
                ps = new ParserState<Info>( new std::ifstream(data_filename) );

                const auto &up = *ptr;
                r = up( *ps );

                if ( r ) 
                {
                        userData = *r;
                        std::string  s = output_result( op_output_mode, userData );
                        std::cout << s << '\n';

                        if ( op_debug )
                                dump_parser_stack( ps, data_filename, "Parsing succeeded" );

                        std::exit( 0 );
                }
                else
                {
                        if ( op_silent ) 
                                std::cout << "Failure parsing data from '" << 
                                        data_filename << "'\n";
                        else
                                dump_parser_stack( ps, data_filename, "Parsing failed" );

                        std::exit( 1 );
                }
        }
        catch ( const std::exception &e )
        {
                std::cout << "Unexpected exception: " << e.what() << '\n';
                std::exit( 1 );
        }
        catch ( const Exception &e )
        {
                std::cout << e.toMessage() << '\n';
                std::exit( 2 );
        }
        catch ( const RuntimeError &e )
        {
                std::cout << "Parser execution error: " << e.what << '\n';
                std::exit( 3 );
        }

        return 0;
}
